### PR TITLE
feat: sticky table of contents with scrollspy for canvas dashboards

### PR DESCRIPTION
**Pull request by rohithreddykota**
Adds a persistent, minimap-style table of contents to canvas dashboards (view surfaces only; not the iframe embed).

- Derives entries from the headings (`h1`–`h3`) rendered inside `markdown` components of the active tab — no new section primitive; stays in sync as content/tabs change via a debounced `MutationObserver`.
- Scrollspy via a single `IntersectionObserver` (no scroll-event math); exactly one active section, with the highlight pinned during click-to-scroll so it doesn't flicker through intermediate sections.
- Collapsed rail of tick bars by default (dots on narrow viewports so they don't overlap content); hover/focus reveals the full list as an overlay flyout that never reflows the dashboard.
- Click smooth-scrolls (respecting `prefers-reduced-motion`) and updates the URL hash; deep-linking on mount is supported. Accessible: `<nav aria-label>`, real links, `aria-current="location"`, focus rings, full heading as the accessible name.

**Checklist:**
- [ ] Covered by tests
- [ ] Ran it and it works as intended
- [ ] Reviewed the diff before requesting a review
- [ ] Checked for unhandled edge cases
- [ ] Linked the issues it closes
- [ ] Checked if the docs need to be updated. If so, create a separate Linear DOCS issue
- [ ] Intend to cherry-pick into the release branch
- [ ] I'm proud of this work!

---
*Developed in collaboration with Claude Code*
image image image

### CODE REVIEW

**Comment by AdityaHegde** [comment]
A couple of UXQA items,
1. When there are 2 markdown components side-by-side,
   1. Clicking on the right one in sidebar will not highlight it, only the left one gets highlighted. 
   2. Scrolling just down will quickly highlight the right one.
   Not sure if there is a legitimate use case for this, so perhaps we can keep the existing behaviour.
2. When the last markdown component is towards the bottom it never gets highlighted. But clicking it does take the user to the bottom so probably not a bad experience?

What do you think @nishantmonu51 ?

File: web-common/src/features/canvas/toc/CanvasTableOfContents.svelte [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+     transparent to pointer events (only `.toc__panel` is interactive), so hovering the empty gutter
+     above/below the rail never holds the flyout open. */
+  .toc {
+    @apply pointer-events-none absolute left-0 top-0 z-40 h-full;
+  }
+
+  /* The rail itself: vertically centered, expanding to the full flyout on hover/focus (which
+     overlays the content to its right rather than reflowing it). `p-2` keeps the first and last bars
+     clear of the rounded corners so they aren't clipped. */
+  .toc__panel {
```
Comment (AdityaHegde) [inline web-common/src/features/canvas/toc/CanvasTableOfContents.svelte]: Lets use `toc-panel`. The naming convention of double underscore is not standard for css. Same applies to double dash classes like `toc__item--active`.

File: web-common/src/features/canvas/toc/CanvasTableOfContents.svelte [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    }
+  }
+
+  function scrollBehavior(): ScrollBehavior {
+    return window.matchMedia?.("(prefers-reduced-motion: reduce)").matches
+      ? "auto"
+      : "smooth";
+  }
+
+  function jumpTo(event: MouseEvent, entry: TocEntry) {
```
Comment (AdityaHegde) [inline web-common/src/features/canvas/toc/CanvasTableOfContents.svelte]: We can move this and watchScrollEnd to a separate class that has logic from loc and scrollspy. Also we should move to svelte5 way for using cross file state. Check web-common/src/lib/store-utils/types.svelte.ts for reference.

**Review by AdityaHegde — CHANGES_REQUESTED** [review]

File: web-common/src/features/canvas/toc/CanvasTableOfContents.svelte (hunk 1 above)
Comment (rohithreddykota) [reply web-common/src/features/canvas/toc/CanvasTableOfContents.svelte]: Done - renamed all classes to kebab-case (toc-panel, toc-item, toc-item-active, toc-bar, toc-text, toc-list, toc-compact) and updated every selector.

File: web-common/src/features/canvas/toc/CanvasTableOfContents.svelte (hunk 2 above)
Comment (rohithreddykota) [reply web-common/src/features/canvas/toc/CanvasTableOfContents.svelte]: Extracted everything into toc-controller.svelte.ts - a CanvasTocController class following the types.svelte.ts convention, with $state entries/activeId as the cross-file state. It absorbs the former scrollspy.ts (IntersectionObserver tracking + lock/unlock) along with refresh, the MutationObserver, jumpTo, and watchScrollEnd; slugify/deriveTocEntries stay as pure helpers in toc.ts. The component is now thin and runes-based, with an $effect owning the controller's lifecycle.

File: web-common/src/features/canvas/toc/toc-controller.svelte.ts [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    event.preventDefault();
+    // Pin the highlight to the target for the duration of the scroll, so it doesn't flicker through
+    // every section the smooth scroll passes; release once the scroll settles.
+    this.locked = true;
+    this.activeId = entry.id;
+    entry.el.scrollIntoView({
+      behavior: this.scrollBehavior(),
+      block: "start",
+    });
+    history.replaceState(history.state, "", `#${entry.id}`);
```
Comment (AdityaHegde) [inline web-common/src/features/canvas/toc/toc-controller.svelte.ts:70]: Do we need replace state here? Using `goto` will instead add support for back and forward navigation buttons. More of a product question @nishantmonu51

File: web-common/src/features/canvas/toc/toc-controller.svelte.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  private refreshTimer: ReturnType<typeof setTimeout> | undefined;
+  private cancelScrollWatch: (() => void) | undefined;
+
+  constructor(private readonly scrollContainer: HTMLElement) {
+    this.refresh();
+
+    // Re-derive on any content change: tab switch, async markdown resolution, edits, filters, or the
+    // `.row-container` appearing (e.g. after required filters are satisfied). Observing the scroll
+    // container rather than `.row-container` keeps this resilient to remounts.
+    this.mutationObserver = new MutationObserver(() => this.scheduleRefresh());
```
Comment (AdityaHegde) [inline web-common/src/features/canvas/toc/toc-controller.svelte.ts]: This is a bit too aggressive. Lets fire events from web-common/src/features/canvas/components/markdown/Markdown.svelte on mount and destroy?

File: web-common/src/features/canvas/toc/toc.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+  // The shallowest level present becomes the top level; anything deeper nests one indent per level.
+  const minLevel = Math.min(...headings.map((h) => h.level));
+
+  const usedIds = new Set<string>();
+
+  return headings.map(({ el, text, level }) => {
+    let id = el.id || slugify(text);
+    // Dedupe collisions with a numeric suffix, e.g. "overview", "overview-2".
+    if (usedIds.has(id)) {
```
Comment (AdityaHegde) [inline web-common/src/features/canvas/toc/toc.ts]: Lets use `getName` from web-common/src/features/entity-management/name-utils.ts . While it is not 1-1 it will suffice to give us a number appended id.

File: web-common/src/features/canvas/toc/toc.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  text: string;
+  /** Nesting depth relative to the shallowest heading present: 0 for top-level, incrementing by
+   *  one per heading level (so h2 and h3 indent differently). */
+  depth: number;
+  /** The heading element itself, used as the scroll target and the IntersectionObserver target. */
+  el: HTMLElement;
+};
+
+/** Turn arbitrary heading text into a URL-hash-safe slug. */
+export function slugify(text: string): string {
```
Comment (AdityaHegde) [inline web-common/src/features/canvas/toc/toc.ts:39]: Lets merge this with web-common/src/lib/string-utils.ts::sanitizeSlug

**Review by AdityaHegde — CHANGES_REQUESTED** [review]

File: web-common/src/features/canvas/toc/toc-controller.svelte.ts (hunk 4 above)
Comment (rohithreddykota) [reply web-common/src/features/canvas/toc/toc-controller.svelte.ts]: Done — replaced the subtree `MutationObserver` with a `CANVAS_TOC_REFRESH_EVENT` that `Markdown.svelte` dispatches on mount, content change, and unmount via a Svelte action; the controller just listens for it, so the TOC no longer re-derives on unrelated chart/table/tooltip updates.

File: web-common/src/features/canvas/toc/toc.ts (hunk 5 above)
Comment (rohithreddykota) [reply web-common/src/features/canvas/toc/toc.ts]: Done — now using `getName` from `name-utils.ts` for the dedup; duplicate anchor ids become `overview_1`, `overview_2`, etc.

File: web-common/src/features/canvas/toc/toc.ts (hunk 6 above)
Comment (rohithreddykota) [reply web-common/src/features/canvas/toc/toc.ts:39]: Done — `slugify` now builds on `sanitizeSlug` for the character replacement, keeping the TOC-specific lowercasing, hyphen-collapsing, and `"section"` fallback.

**Review by AdityaHegde — APPROVED** [review]
Approving from code side. @nishantmonu51 please do a UXQA once

**Review by nishantmonu51 — COMMENTED** [review]
The branch no longer merges: `main` rewrote `CanvasDashboardWrapper.svelte` to Svelte 5 runes in #9746, and `git merge-tree` reports a content conflict in that file, so `export let showTableOfContents` and the `bandWidth`/`scrollContainer` bindings will need porting to the runes form on rebase.

Two smaller points:

`deepLinked` latches on the first derive that yields any entries, but templated markdown initially renders its raw content — `resolvedContent` falls back to `content` before the resolve query returns. A heading like `## Revenue {{ ... }}` therefore produces id `revenue-...` on the first pass and its resolved id only later, so a deep link to the resolved id is never honored because the latch has already flipped (`toc-controller.svelte.ts:100`, `components/markdown/Markdown.svelte:82`).

`web-admin/src/features/scheduled-reports/export/CanvasPdfReportExport.svelte:170` mounts `CanvasDashboardEmbed` with the default `tableOfContents = true`, so the TOC controller and its IntersectionObserver run inside the off-screen capture host for no benefit. Pass `tableOfContents={false}` as the embed route does.

File: web-common/src/features/canvas/toc/toc-controller.svelte.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    const rowContainer =
+      this.scrollContainer.querySelector<HTMLElement>(".row-container");
+    this.entries = deriveTocEntries(rowContainer);
+    this.observe();
+
+    // On first render, honor a deep link to a section (tab selection is handled separately via the
+    // existing `tabs` URL param). Use an instant jump so the page doesn't animate on load.
+    if (!this.deepLinked && this.entries.length > 0) {
+      this.deepLinked = true;
+      const hash = decodeURIComponent(window.location.hash.slice(1));
```
Comment (nishantmonu51) [inline web-common/src/features/canvas/toc/toc-controller.svelte.ts:102]: The hash read here is usually gone by the time `refresh` runs. `CanvasInitialization.svelte:122` calls `canvasEntity.onUrlChange` as soon as the store resolves, and for a non-isolated canvas `handleCanvasRedirect` does ``goto(`?${snapshotSearchParams}`, { replaceState: true })`` (and the equivalent for a home bookmark or for default params) whenever `searchParams.size === 0` — see `stores/canvas-entity.ts`, `handleCanvasRedirect`. Resolving a relative `?...` URL against the current location discards the fragment, so `/canvas/foo#overview` becomes `/canvas/foo?tr=...` before, or racing with, the first `refresh`. Any canvas with a last-visited snapshot, a home bookmark, or default params therefore never honors the deep link, and even when the scroll wins the race the hash is stripped from the URL. Either carry the hash through the redirect or capture it once before `onUrlChange` runs.

File: web-common/src/features/canvas/toc/toc.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  if (headings.length === 0) return [];
+
+  // The shallowest level present becomes the top level; anything deeper nests one indent per level.
+  const minLevel = Math.min(...headings.map((h) => h.level));
+
+  const usedIds: string[] = [];
+
+  return headings.map(({ el, text, level }) => {
+    // Dedupe collisions with a numeric suffix, e.g. "overview", "overview_1".
+    const id = getName(el.id || slugify(text), usedIds);
```
Comment (nishantmonu51) [inline web-common/src/features/canvas/toc/toc.ts:78]: `getName` increments a trailing number (`INCREMENT = /(\d+)$/`, applied as `result.replace(INCREMENT, (m) => (+m + 1).toString())`) rather than appending a suffix, so headings `2024`, `2024`, `2025` are assigned ids `2024`, `2025`, `2026` — `#2025` then scrolls to the second `2024` heading instead of the `2025` one, and a duplicate `q1-2025` is likewise renamed to `q1-2026`. `toc.spec.ts` only covers the alphabetic case (`overview_1`), which is why this passes. A local `-1`/`-2` suffix loop avoids borrowing the resource-name helper's semantics.
